test/format/replay.c

```diff
@@ -0,0 +1,543 @@
+     * the intended stop timestamp. We pick a stop timestamp far enough in the future that it's
+     * rather unlikely to happen.
+     */
+    WT_PUBLISH(g.stop_timestamp, g.timestamp + 0x10000);
+}
+
+/*
+ * replay_maximum_committed --
+ *     For predictable replay runs, return the largest timestamp that's no longer in use.
+ */
+uint64_t
+replay_maximum_committed(void)
+{
+    uint64_t commit_ts, ts;
+    uint32_t lane;
+
+    /*
+     * The calculation is expensive, and does not need to be accurate all the time, and it's okay to
+     * be behind. So we use a cached value most of the time.
+     */
+    ts = g.replay_cached_committed;
+    if (ts == 0 || __wt_atomic_addv32(&g.replay_calculate_committed, 1) % 20 == 0) {
+        WT_ORDERED_READ(ts, g.timestamp);
+        testutil_check(pthread_rwlock_wrlock(&g.lane_lock));
+        for (lane = 0; lane < LANE_COUNT; ++lane) {
+            if (g.lanes[lane].in_use) {
+                commit_ts = g.lanes[lane].last_commit_ts;
+                if (commit_ts != 0)
+                    ts = WT_MIN(ts, commit_ts);
+            }
+        }
+        if (ts == 0)
+            ts = 1;
+        g.replay_cached_committed = ts;
+        testutil_check(pthread_rwlock_unlock(&g.lane_lock));
+    }
+    return (ts);
+}
+
+/*
+ * replay_operation_enabled --
+ *     Return whether an operation type should be enabled in the configuration.
+ */
+bool
+replay_operation_enabled(thread_op op)
+{
+    if (!GV(RUNS_PREDICTABLE_REPLAY))
+        return (true);
+
+    /*
+     * We don't permit truncate operations with predictable replay.
+     *
+     * Currently, we use an operation's timestamp to help derive the operation's key.
+     * The last N bits of the timestamp are used as the last bits of the key (where
+     * 2^N == LANE_COUNT). These last N bits give the lane number, and within each
+     * lane we track the progress of operations for that lane. Using lanes, we can
+     * track and guarantee that only a single operation is active in a lane at once,
+     * and therefore we can't have multiple operations on a single key performed out
+     * of order or simultaneously. The truncate operation, for a small set of keys,
+     * would reserve multiple consecutive lanes (probably okay) and for larger sets,
+     * would reserve the entire set of lanes. This would effectively require all
+     * threads to get into a holding state, waiting for the truncate to start and then
+     * complete before continuing with their next operation. While we could fudge this
+     * in certain ways (e.g. operations with 10000 timestamps of a truncate would be
+     * forced to stay out of its table), there still would be a lot of details, and
+     * some rethink of our lane strategy. Even getting this to work, we would have
+     * a truncate that had the whole table to itself, which doesn't seem like an
+     * effective test.
+     */
+    if (op == TRUNCATE)
+        return (false);
```

Comment:
Alphabetize TRUNCATE and its comment below REMOVE

Reply:
Fixed.